WT-10253 Make sure session dhandle sweep and session cursor sweeps keep up

ddanderson opened the pull request.

sueloverso commented on src/include/session.h:
```diff
@@ -82,8 +83,8 @@ struct __wt_session_impl {
-     * of the data handle's URI. The hash table list is kept in allocated memory that lives across
-     * session close - so it is declared further down.
+     * of the data handle's URI. The hash table list is kept in allocated memory until a session
+     * is closed, when it is freed. It is declared further down.
```
This comment change is not correct, I think. Everything in the session after `flags` is not freed on close. Everything in the session structure after `WT_SESSION_CLEAR_SIZE` persists after a close. That was the specific point of the original form of the comment.

ddanderson replied:
Right - I'll clarify the comment, the elements are discarded and freed even though the hash array is kept.

ddanderson replied:
Fixed.

sueloverso commented on src/include/session.h:
```diff
@@ -94,7 +95,8 @@ struct __wt_session_impl {
     uint32_t cursor_sweep_countdown; /* Countdown to cursor sweep */
-    uint64_t last_cursor_sweep;      /* Last sweep for dead cursors */
+    uint64_t last_cursor_sweep;      /* Last regular sweep for dead cursors */
+    uint64_t last_cursor_big_sweep;  /* Last big sweep for dead cursors */
```
Swap for alphabetizing.

etienneptl replied:
Should we take the opportunity to swap the `cursor_sweep` ones as well?

sueloverso commented on src/session/session_api.c:
```diff
@@ -52,29 +52,58 @@ __wt_session_reset_cursors(WT_SESSION_IMPL *session, bool free_buffers)
      * Periodically sweep for dead cursors; if we've swept recently, don't do it again.
+     *
+     * Each call of this sweep function visits all the cursors in some number of buckets used by the
+     * cursor cache. This determines if any of the visited cursors reference dead or dying data
```
"This determines" -> what exactly is "this" referring to? The call of the sweep function? The visiting of some number of buckets? The actual sweeping?

ddanderson replied:
Fixed

sueloverso commented on src/session/session_api.c:
```diff
@@ -52,29 +52,58 @@ __wt_session_reset_cursors(WT_SESSION_IMPL *session, bool free_buffers)
+     * sessions give up their references. So sweeping the cursor cache (for all sessions!) is a
+     * prerequisite for the connection data handle sweep to find handles that can be freed.
+     *
+     * So how many buckets to visit? This function is called in two ways. When a session reset is
```
Instead of a rhetorical question, I suggest something like:
We determine the number of buckets to visit based on how the function is called....

ddanderson replied:
Fixed

sueloverso commented on src/session/session_api.c:
```diff
@@ -52,29 +52,58 @@ __wt_session_reset_cursors(WT_SESSION_IMPL *session, bool free_buffers)
+     *
+     * Each call of this sweep function visits all the cursors in some number of buckets used by the
+     * cursor cache. This determines if any of the visited cursors reference dead or dying data
+     * handles and should be fully closed and evicted from the cache. Evicting a cursor from the
```
If possible I'd suggest finding different words instead of `evicted from the cache` because that could be confusing for someone thinking you're referring to the data cache and eviction threads, etc.

ddanderson replied:
Fixed, now: `removed from the cache`

sueloverso commented on src/session/session_api.c:
```diff
@@ -52,29 +52,58 @@ __wt_session_reset_cursors(WT_SESSION_IMPL *session, bool free_buffers)
+     *
+     * So how many buckets to visit? This function is called in two ways. When a session reset is
+     * done, this is called with big_sweep being true. We are not in a performant path when doing a
+     * reset, so if enough time has passed, we'll walk through a minium of a quarter of the buckets,
```
typo: minimum

ddanderson replied:
Fixed

sueloverso commented on src/session/session_dhandle.c:
```diff
@@ -733,7 +733,7 @@ __session_get_dhandle(WT_SESSION_IMPL *session, const char *uri, const char *che
 
     /* Sweep the handle list to remove any dead handles. */
-    __session_dhandle_sweep(session);
+    __wt_session_dhandle_sweep(session);
```
Not related to this change but it is the end of the change: I think the test you wrote should also be added to this.

sueloverso reviewed (commented):
I have several comments. Mostly looks good but I'd like to see the test too.

ddanderson commented:
@keitharnoldsmith and @sueloverso, I intend to include a test, but that's a little involved, so I thought you might want a head start on the core part of change, this is all ready for review.

ddanderson commented:
@sueloverso, @etienneptl @keitharnoldsmith I think I've addressed your comments, and I finally have a reasonable test case (I think).  The test takes a long time to run so it's marked as "long" so isn't part of the usual pull request.  It passes for me locally, and I just added the appropriate tasks to the PR test so it running now.

sueloverso commented on src/session/session_api.c:
```diff
@@ -52,43 +52,72 @@ __wt_session_reset_cursors(WT_SESSION_IMPL *session, bool free_buffers)
+     * connection data handle sweep to find handles that can be freed.
+     *
+     * We determine the number of buckets to visit based on how this function is called. When a
+     * session reset is done, this function is called with big_sweep set to true. We are not in a
```
I'd prefer some form of this comment be up in session reset. This function's comment should only say what it does when big_sweep is true rather than itemize who calls it in what form. That higher level code may change and this function would not know that. Maybe something like:

We determine the number of buckets to visit based on how this function is called. When big_sweep is set to true
and enough time has passed, walk through ...

sueloverso commented on src/session/session_api.c:
```diff
@@ -52,43 +52,72 @@ __wt_session_reset_cursors(WT_SESSION_IMPL *session, bool free_buffers)
+     * minimum of a quarter of the buckets, and as long as we're making progress finding enough
+     * cursors to close, we'll continue on, up to the entire set of buckets.
+     *
+     * When we're called in a fast path (releasing a closed cursor to the cache), big_sweep is
```
Again, I'd remove the part about the caller and just say "When big_sweep is false we start with a small..."

Adding a comment up in the caller to say, "This is a performant path so don't do a big sweep on this call."

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# test_sweep04.py
+# Test lots of tables with more steadily created and dropped.
```
"with more tables steadily..."
Also a single sentence about what the test is checking for:
Test that the count of dhandles doesn't exceed...

ddanderson replied:
Fixed

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+    transient_tables = 100         # Number of transient tables at any time
+    transient_examined = 10        # Number of transient tables opened at a time
+    ratio_examined = 0.01          # The chance that transient tables are examined
+    transient_table_max = 10000    # defines the length of the run
```
minor: Capitalize "defines"

ddanderson replied:
Fixed

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+
+    # Configuration values for the run. If any of these values are changed,
+    # the simulation will run, but the acceptance criteria may fail at the end.
+    core_tables = 10               # Number of core tables that always exist
```
End all of the comments here with periods. Some do, some don't.

ddanderson replied:
Fixed

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+    numkv = 1                      # Number of k/v pairs. Shouldn't matter for this test.
+    nsessions = 100                # Number of sessions in our pool.
+
+    conn_config = 'file_manager=(close_handle_minimum=0,' + \
```
Since this is a sweep test, the values in here probably matter. A comment describing them and their impact would be helpful.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+        for i in range(0, count):
+            c = session.open_cursor(uri_maker(start + i))
+            self.assertEquals(c[1], 1)
+            c.close()
```
All of these small functions could use a one-sentence description/comment.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+        for i in range(0, self.core_tables):
+            self.create_table(self.core_uri(i))
+
+        created_transient = 0    # running total of transients created
```
Take a pass through all the comments to start with capitals and end with periods.

sueloverso replied:
Can you do this for all the one-line comments that are complete sentences?

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+                thistime = time.time()
+                delta = thistime - lasttime
+                if delta < 3.0:
+                    time.sleep(3.0 - delta)
```
Could this '3.0' and the '3' in the config above be defined at the top so if we ever want to adjust to 2 or 4 seconds, it would be possibly two lines next to each other.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+            else:
+                # look at a core table
+                t = r.rand_range(0, self.core_tables)
+                self.examine(rand_session, self.core_uri, t, 1)
```
What is 't'? The comment says we're looking at a single table, and then I thought 't' indicated a range of tables, but then I see that the '1' as the last arg means one table. So I'm not sure what 't' means (in either call).

ddanderson replied:
I've added comments here and other functions, hopefully it's clearer.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+                t = r.rand_range(0, self.core_tables)
+                self.examine(rand_session, self.core_uri, t, 1)
+
+            if loopcount % 100 == 99:
```
Add a comment to indicate we're only gathering and processing the stats now, not checking them.

ddanderson replied:
Added.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+
+        # Half the run is dhandle growth, and the second half should see a decline.
+        # If everything is working right in the first half of the run, we should start
+        # to see the number of dhandles start to asymptote. So we check the slope of
```
'asymptote' is a noun only, not a verb (according to dictionary.com).

ddanderson replied:
Yes, fixed!

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+        # Note, we don't check the first half average, it's likely to be big, but its size
+        # depends on many factors. The important thing is that the slope has flattened out.
+        self.assertLess(abs(q2_slope), q1_slope)
+        self.assertLess(abs(q2_slope), 15.0)
```
Add to the comment "The important thing..." about how 15.0 means it has flattened out.

ddanderson replied:
Added.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,225 @@
+        self.assertLess(abs(q2_slope), 15.0)
+
+        # At the end of the run, we expect a pretty flat slope and a pretty small number of dhandles.
+        self.assertLess(abs(end_run_slope), 5.0)
```
Same here about the meaning of 5.0.

ddanderson replied:
Added.

sueloverso reviewed (approved):
LGTM with a bunch of comments on comment wording and such. I think the code looks good. I'll take another look when changes are made but don't want to hold up a merge either.

sueloverso commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,248 @@
+# Test that the total number of dhandles, while increasing,
+# starts to level off, that is, the sweeps are keeping up.
+# Then test that if we only access the core tables for a while,
+# the total number of dhandles comes back done to a small number.
```
Minor: 'done' should be 'down'

sueloverso reviewed (approved):
LGTM with a few very minor text edits.

keitharnoldsmith commented on test/suite/test_sweep04.py:
```diff
@@ -0,0 +1,248 @@
+                self.pr('  file_open={}'.format(files_open))
+                dhandle_counts.append(dhandles)
+                stat_cursor.close()
+                if loopcount % 100 == 999:
```
Typo? `loopcount % 100` will never be `999`

ddanderson replied:
Thanks, I'll fix!

ddanderson replied:
Ha, 'fixing' it causes the test to fail, as it writes to stdout, and that fails the test.  So I fixed, but disabled with a comment.

keitharnoldsmith replied:
👍

keitharnoldsmith reviewed (approved):
LGTM. One minor comment.
